### ISSUE TITLE
Multiprocessing data loader

### ISSUE THREAD

**Issue by vafl** [awslabs/gluon-ts#682]
## Description
We currently have a bottleneck in our transformation/data loading. For fast models we starve the GPU, because our transformations are not fast enough. By just returning a pre-computed batch again and again I estimated that we could get a speedup in training of at least 2x for most models.

Fixing this should be straight forward.
Mxnet [supports multiprocessing data loading](https://github.com/apache/incubator-mxnet/blob/f619c52dfc3787e3097b37ba28f7065bca910b97/python/mxnet/gluon/data/dataloader.py#L438)  for the standard mxnet datasets. We cannot use this directly with our current design, because our feature pipeline is  based on iterators. However there are examples of how to do this with iterable datasets aswell (e.g. [pytorch](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset)).

**Comment by jaheba** [comment]
I think a first step would be to actually benchmark our current implementation, and thus create a baseline we can compare against. 

Maybe we can even get more fine-grained numbers, differentiating between loading of json-data and applying transformations.

Maybe these reveal some bottlenecks we are currently not aware of.

**Comment by AaronSpieler** [comment]
I have done that experiment and it seems almost all the time is spent on fetching the data (or at least not doing the transformation):

```
transformation = estimator.create_transformation()
transformation.estimate(iter(dataset.train))

training_data_loader = TrainDataLoader(
            dataset=dataset.train,
            transform=transformation,
            batch_size=trainer.batch_size,
            num_batches_per_epoch=trainer.num_batches_per_epoch,
            ctx=trainer.ctx
            #,dtype=self.dtype,
        )
list(training_data_loader); // caches the transformation in case of measuring cached
```

```
import timeit
num_iter=100
print(f"Runtime: {timeit.timeit(lambda: list(training_data_loader), number=num_iter)/num_iter}")
```
Originally: `0.5170380667998689`
Cached: `0.5077913579298183` // this does not compute the transformations


caching:
```
class Chain(Transformation):
    """
    Chain multiple transformations together.
    """

    cached = False
    cache = None

    @validated()
    def __init__(self, trans: List[Transformation]) -> None:
        self.transformations = []
        for transformation in trans:
            # flatten chains
            if isinstance(transformation, Chain):
                self.transformations.extend(transformation.transformations)
            else:
                self.transformations.append(transformation)

    def __call__(
        self, data_it: Iterable[DataEntry], is_train: bool
    ) -> Iterator[DataEntry]:
        tmp = data_it
        if not self.cached:
            for t in self.transformations:
                tmp = t(tmp, is_train)
            self.cached = True
            self.cache = tmp
            print("Just cached")
        else:
            return self.cache
        return tmp

    def estimate(self, data_it: Iterator[DataEntry]) -> Iterator[DataEntry]:
        return reduce(
            lambda x, y: y.estimate(x), self.transformations, data_it
        )
```

**Comment by jaheba** [comment]
Having some iterations per second number of different parts would be valuable. AFAIK we had something like that in the beginning. 

We could also think about just loading all data into memory. This has some implications to transformations (ie shuffling), but probably an easy speed up. 

It would be interesting to know how fast the different components are:

* reading of a line
- use of json loads
- conversion into numpy arrays and pandas timestamps

We could also think about supporting different file formats (parquet).

Thus, what I want is to have the clearest picture possible.

**Comment by AaronSpieler** [comment]
> Having some iterations per second number of different parts would be valuable. AFAIK we had something like that in the beginning.
> 
> We could also think about just loading all data into memory. This has some implications to transformations (ie shuffling), but probably an easy speed up.
> 
> It would be interesting to know how fast the different components are:
> 
> * reading of a line
> 
> * use of json loads
> * conversion into numpy arrays and pandas timestamps
> 
> We could also think about supporting different file formats (parquet).
> 
> Thus, what I want is to have the clearest picture possible.

So I have tried alternatives to `json_loads` in particular `linecache.getline`, but there is no noticeable difference, and there are a couple of reasons for that:

In particular I found out that:
* we are already always caching the whole dataset as far as python caching is concerned, meaning that no entry is read twice from the file, independent of how often it is accessed, however, its not the transformed dataset that is cached
** this also means that maybe only `memory pinning` could help here, which im not sure is possible in a sensible way for our datasets
** or really the conversion of the datasets into the right format costs so much time, which could be parallelised
* due to the nature of our dataset, being variable length and dicts, there is no way to skip to a particular line in the file, without reading the previous lines, so efficiently reading parts of the datasets is not possible, however, the dataset can be read and then split among workers explicitly

**Comment by AaronSpieler** [comment]
Thus overall I agree with jaheba, maybe it would be useful to see what exactly the bottleneck is. Maybe a notebook as a minimal reproducible example would help a lot, since for example on my laptop im not able to reproduce these kinds of speedups, but with the right notebook on a SagemakerNotebook instance this should be no problem. I have a sunspicion that the actusl bottleneck is neither the loading of tge data into RAM nor the transformstion, but ratger the GPU fetching the data out of RAM, which could only be adressed with memory pinning or a GPU memory based queue.

**Comment by vafl** [comment]
Here is an extreme example using wavenet:

https://gist.github.com/vafl/9fec2eef89135d9beb54702a5d6cbbf9

If you run our normal transformation chain one epoch takes around 70s. If you return the same batch again and again, training takes 5s. 

To make sure this is not caching related I actually return the same batch in both cases, but run the transformation in one of them and not in the other.

**Comment by AaronSpieler** [comment]
Thx @vafl , with this I was able to reproduce the performance gains you were talking about.

**Comment by AaronSpieler** [comment]
The pull #689 fixed this.
